dnsapi/dns_mydevil.sh

```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+
+# MyDevil.net API (2019-02-03)
+#
+# MyDevil.net already supports automatic Let's Encrypt certificates,
+# except for wildcard domains.
+#
+# This script depends on `devil dns` that MyDevil.net provides,
+# which means that it works only on server side.
+#
+# Author: Marcin Konicki <https://ahwayakchih.neoni.net>
+#
+########  Public functions #####################
+
+#Usage: dns_mydevil_add   _acme-challenge.www.domain.com   "XKrxpRBosdIKFzxW_CT3KLZNf6q0HG9i01zxXp5CPBs"
+dns_mydevil_add() {
+  local fulldomain=$1
+  local txtvalue=$2
+  local domain=""
+
+  _info "Using mydevil"
+
+  domain=$(mydevil_get_domain "$fulldomain")
+  if ! mydevil_check_record "$fulldomain"; then
+    _err "Invalid record name: does not start with '_acme-challenge'."
+    return 1
+  fi
+
+  if [ -z "$domain" ]; then
+    _err "Invalid domain name: could not find root domain of $fulldomain."
+    return 1
+  fi
+
+  _info "Adding $fulldomain record for domain $domain"
+  if devil dns add "$domain" "$fulldomain" TXT "$txtvalue"; then
+    _info "Successfully added TXT record, ready for validation."
+    return 0
+  else
+    _err "Unable to add DNS record."
+    return 1
+  fi
+}
+
+#Usage: fulldomain txtvalue
+#Remove the txt record after validation.
+dns_mydevil_rm() {
+  local fulldomain=$1
+  local txtvalue=$2
+  local domain=""
+
+  _info "Using mydevil"
+
+  domain=$(mydevil_get_domain "$fulldomain")
+  if ! mydevil_check_record "$fulldomain"; then
+    _err "Invalid record name: does not start with '_acme-challenge'."
+    return 1
+  fi
+
+  if [ -z "$domain" ]; then
+    _err "Invalid domain name: could not find root domain of $fulldomain."
+    return 1
+  fi
+
+  for id in $(devil dns list "$domain" | grep "$fulldomain" | awk '{print $1}'); do
+    _info "Removing record $id from domain $domain"
+    devil dns del "$domain" "$id" || _err "Could not remove DNS record."
+  done
+}
+
+####################  Private functions below ##################################
+
+# Usage: mydevil_check_record "_acme-challenge.www.domain.com" || _err "Invalid record name"
+mydevil_check_record() {
+  local record=$1
```

Comment:
you can not use `_acme-challenge.` prefix for subdomain and domain.
we have  dns alias mode: https://github.com/Neilpang/acme.sh/wiki/DNS-alias-mode

The full domain you get here is not necessarily starting with `_acme-challenge.`

It's recommended to `list` all he zones by `devil` and compare with each one to check which part is root domain and which is subomain.

see:  https://github.com/Neilpang/acme.sh/wiki/DNS-alias-mode

Reply:
This function is only to make sure that the record is valid, and not some real, non-acme related record that should not be touched.

To get the root domain name, another function is used.

Is there a variable that keeps the prefix part of the name, that i could check?

Reply:
@Neilpang i removed whole function and changed code so it does not depend on `_acme-challenge` prefix at all. Instead it makes sure to only remove `TXT` records that match both `txtdomain` and `txtvalue`.